src/compiler/checker.ts

```diff
@@ -21248,11 +21248,15 @@ namespace ts {
                 type;
         }
 
-        function getWidenedLiteralLikeTypeForContextualType(type: Type, contextualType: Type | undefined) {
-            if (!isLiteralOfContextualType(type, contextualType)) {
-                type = getWidenedUniqueESSymbolType(getWidenedLiteralType(type));
+        function getWidenedLiteralLikeTypeForContextualType(type: Type, contextualType: Type | undefined, node?: Node) {
+            if (isLiteralOfContextualType(type, contextualType)) {
+                return type;
             }
-            return type;
+            const instantiatedContextualType = node && instantiateContextualType(contextualType, node) || contextualType;
+            if (instantiatedContextualType !== contextualType && isLiteralOfContextualType(type, instantiatedContextualType)) {
+                return type;
+            }
+            return getWidenedUniqueESSymbolType(getWidenedLiteralType(type));

```

Comment:
This change, allowing us to check if `type` is a literal of either the uninstantiated or instantiated contextual type, was necessary to fix this case:

```ts
declare function pick<O, T extends keyof O>(keys: T[], obj?: O): Pick<O, T>;
const _    = pick(['b'], { a: 'a', b: 'b' }); // T: "b"
const {  } = pick(['b'], { a: 'a', b: 'b' }); // T: "b" | "a" ??? (before fix)
```

When the binding pattern is `{}`, we infer `never` for `T` and use that to instantiate the contextual type `T[]` → `never[]` for the argument `['b']`. Because `"b"` is not a “literal of contextual type `never`,” we widen to `string`, which eventually gets replaced with the constraint `keyof O` → `"b" | "a"`.

The change here checks to see if `"b"` is a literal of contextual type `T extends keyof O`, which is already handled by substituting the base constraint, which allows us to see that we should not widen from the literal `"b"`.

Reply:
As a side note, the reason fixing this is important even given the obvious uselessness of the construct `const {} = anything` is that we ought to be able to deliver accurate completions inside the `{}`. Prior to this fix, you got both `a` and `b` since that’s what `T` had in its domain, but as soon as you accept the `a` completion, the contextual type changes and therefore the inference changes and it becomes an error, which is a terrible experience.